### ISSUE TITLE
SQL auditing not re-enabled after manually being disabled

### ISSUE THREAD

**Issue by g-psantos** [hashicorp/terraform-provider-azurerm#14859]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v1.0.11
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.90.0
```

### Affected Resource(s)

- `azurerm_mssql_server_extended_auditing_policy`
- `azurerm_mssql_database_extended_auditing_policy`
- `azurerm_monitor_diagnostic_setting`

### Terraform Configuration Files

I'm using the following code, which is based on the [example](https://github.com/hashicorp/terraform-provider-azurerm/tree/main/examples/sql-azure/sql_auditing_log_analytics) that's linked to in the documentation for logging to a Log Analytics workspace.

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "2.90.0"
    }
  }
}

provider "azurerm" {
  features {}
}

resource "azurerm_mssql_server" "this" {
  name                         = "your-db-server-name"
  resource_group_name          = "your-rg-name"
  location                     = "eastus"
  version                      = "12.0"
  administrator_login          = "administrator"
  administrator_login_password = "thisIsDog1!"
  minimum_tls_version          = "1.2"
}

resource "azurerm_mssql_database" "this" {
  server_id = azurerm_mssql_server.this.id
  name      = "test-db"
}

resource "azurerm_log_analytics_workspace" "this" {
  name                = "your-workspace-name"
  location            = "eastus"
  resource_group_name = "your-rg-name"
  sku                 = "PerGB2018"
  retention_in_days   = 30
}

resource "azurerm_monitor_diagnostic_setting" "this" {
  name                       = "DefaultAuditingSettings"
  target_resource_id         = format("%s/databases/master", azurerm_mssql_server.this.id)
  log_analytics_workspace_id = azurerm_log_analytics_workspace.this.id

  # NOTE: All `log` and `metric` categories need to be defined or Terraform
  #       will be in eternal drift trying to change falses and zeros to nulls
  dynamic "log" {
    for_each = local.database_logs
    content {
      category = log.key
      enabled  = log.value
      retention_policy {
        enabled = false
        days    = 0
      }
    }
  }

  dynamic "metric" {
    for_each = local.database_metrics
    content {
      category = metric.key
      enabled  = metric.value
      retention_policy {
        enabled = false
        days    = 0
      }
    }
  }
}

locals {
  database_logs = {
    "SQLSecurityAuditEvents"      = true,
    "DevOpsOperationsAudit"       = true,
    "AutomaticTuning"             = false,
    "Blocks"                      = false,
    "DatabaseWaitStatistics"      = false,
    "Deadlocks"                   = false,
    "Errors"                      = false,
    "QueryStoreRuntimeStatistics" = false,
    "QueryStoreWaitStatistics"    = false,
    "SQLInsights"                 = false,
    "Timeouts"                    = false,
  }

  database_metrics = {
    "Basic"                  = false,
    "InstanceAndAppAdvanced" = false,
    "WorkloadManagement"     = false,
  }
}

resource "azurerm_mssql_database_extended_auditing_policy" "this" {
  database_id            = "${azurerm_mssql_server.this.id}/databases/master"
  log_monitoring_enabled = true
}

resource "azurerm_mssql_server_extended_auditing_policy" "this" {
  server_id              = azurerm_mssql_server.this.id
  log_monitoring_enabled = true
}
```

### Expected Behaviour

When the server is first deployed, it should be configured with auditing to a Log Analytics workspace. Subsequently, if the auditing settings are changed via Azure Portal, CLI, or PowerShell, Terraform should detect those changes and restore auditing settings to their intended state.

(In this specific example, Microsoft DevOps should also be audited, but that's a separate issue that's already being tracked in #12416.)

### Actual Behaviour

On first deployment, auditing settings are configured correctly (save for the DevOps logging category, but ignore that). If you then go to the Azure Portal and disable auditing, Terraform detects that something's changed and proposes to fix it, as per the plan below.

```hcl
Terraform used the selected providers to generate the following execution plan.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place

Terraform will perform the following actions:

  # azurerm_monitor_diagnostic_setting.this will be created
  + resource "azurerm_monitor_diagnostic_setting" "this" {
      + id                         = (known after apply)
      + log_analytics_workspace_id = "***"
      + name                       = "DefaultAuditingSettings"
      + target_resource_id         = "***/databases/master"

      # `log` and `metric` blocks omitted here, but they're as expected given
      # the Terraform configuration pasted above
    }

  # azurerm_mssql_database_extended_auditing_policy.this will be updated in-place
  ~ resource "azurerm_mssql_database_extended_auditing_policy" "this" {
        id                                      = "***/databases/master/extendedAuditingSettings/Default"
      ~ log_monitoring_enabled                  = false -> true
        # (3 unchanged attributes hidden)
    }

Plan: 1 to add, 1 to change, 0 to destroy.
```

**However,** if you then go back into the Azure Portal you can see that auditing is _not_ re-enabled (and you can further confirm this by logging into the test database and looking at the logs in the Log Analytics workspace -- nothing should be there).

### Steps to Reproduce

1. `terraform apply`
2. Confirm that auditing is properly configured by going into the Azure Portal
3. Disable auditing in the Azure Portal
4. `terraform apply` again (you'll see something along the lines of the plan above, suggesting that auditing will be re-enabled)
5. Go back to the Azure Portal -- auditing should be disabled still

**Comment by sinbai** [comment]
@g-psantos thank you for opening this issue. 

After step 3 (Disable auditing in the Azure Portal), check the  value of "isAzureMonitorTargetEnabled" (map the log_monitoring_enabled property in terraform) through [API](https://docs.microsoft.com/en-us/rest/api/sql/2020-08-01-preview/extended-server-blob-auditing-policies/get#code-try-0), you would find that its value is still true), which causes terraform to not detect the log_monitoring_enabled of the azurerm_mssql_server_extended_auditing_policy is changed, but only got log_monitoring_enabled of azurerm_mssql_database_extended_auditing_policy is changed. However, enabling the log monitoring of database can not enable server log monitoring at the same time. So, as you can see the auditing is still disabled in the Azure Portal.

If  I replace the step3 with calling [API](https://docs.microsoft.com/en-us/rest/api/sql/2020-08-01-preview/extended-server-blob-auditing-policies/create-or-update#code-try-0) to disable log monitoring, everything is fine. When running terraform apply in step 4 again, you would got the log_monitoring_enabled of the azurerm_mssql_server_extended_auditing_policy would be detected as follows:
 ```
# azurerm_mssql_database_extended_auditing_policy.this will be updated in-place
  ~ resource "azurerm_mssql_database_extended_auditing_policy" "this" {
        id                                      = "/subscriptions/XXX/resourceGroups/acctestRG-elena11/providers/Microsoft.Sql/servers/db-server-02/databases/master/extendedAuditingSettings/Default"
      ~ log_monitoring_enabled                  = false -> true
        # (3 unchanged attributes hidden)
    }

  # azurerm_mssql_server_extended_auditing_policy.this will be updated in-place
  ~ resource "azurerm_mssql_server_extended_auditing_policy" "this" {
        id                                      = "/subscriptions/XXX/resourceGroups/acctestRG-elena11/providers/Microsoft.Sql/servers/db-server-02/extendedAuditingSettings/Default"
      ~ log_monitoring_enabled                  = false -> true
        # (3 unchanged attributes hidden)
    }
```

**Comment by g-psantos** [comment]
@sinbai Thank you for looking into this and providing an explanation of what's going wrong behind the scenes. Our concern is that someone might disable audit logging on the server via the Portal, and even though we'd see _some_ drift in Terraform, audit logging wouldn't actually be re-enabled with a subsequent deployment.

The API references a `state` property, which can be `Enabled` or `Disabled` and seems to be what actually controls whether auditing is in effect. This property is also available in the `DatabaseBlobAuditingPolicyProperties struct` of the Azure SDK for Go (at least in the `master` branch). Would it be possible to update the AzureRM provider to expose that property? (If you or someone else decides to make that change, it would be nice to also have `AuditActionsAndGroups` available.)

**Comment by sinbai** [comment]
@g-psantos thanks for your response. For the exposing the `state` property, disabling auditing means delete the policy. That means when the azurerm_mssql_server_extended_auditing_policy below is removed, the `state` would be set to `Disabled` in terraform.

```
resource "azurerm_mssql_database_extended_auditing_policy" "this" {
  database_id            = "${azurerm_mssql_server.this.id}/databases/master"
  log_monitoring_enabled = true
}
```

**Comment by g-psantos** [comment]
@sinbai The issue is that when audit logging is disabled via the Azure Portal, they seem to set `state = Disabled` on the server audit policy. However, Terraform fails to detect that and, consequently, also fails to re-enable logging, leaving operators oblivious to a pretty serious drift.

So, if conceptually it does not make sense to expose `state` via the `azurerm_mssql_server_extended_auditing_policy` resource, then shouldn't the provider at least always check that `state` is set to `Enabled` with every refresh?

**Comment by katbyte** [comment]
closed by #15624

**Comment by sinbai** [comment]
@g-psantos thanks for your feedback. I have submitted PR to support the `enabled` property in terraform. That would be released in the incoming release 2.99 so please track that to see whether you would then be unblocked.

**Comment by g-psantos** [comment]
Confirmed that release 2.99 addresses the problem in this issue. Thank you for fixing!
